### PR TITLE
Fix endless loading timeline due to conflicting chunks

### PR DESCRIPTION
**Pull request by SpiritCroc**
I ran into following scenario recently: at some point, scrolling down to more recent messages wasn't working anymore, it would endlessly show the loading animation. This was reproducible, so I started debugging. It turned out:

- The TimelineChunk I was viewing had `chunkEntity.nextChunk == null`.
- When trying to download a new chunk, `TokenChunkEventPersistor` aborted, as the next chunk was already in the database.

It turned out that I had multiple chunks in my DB, so only one had the correct reference to `nextChunk`, while the other one, which I was viewing, was stuck at loading forever.

This commit works around this by "repairing" links in this case. I assume there is a better way/place to fix it, but this works for me for now. I figured it might still be interesting for you, maybe you have some suggestions how to better fix it.

Signed-off-by: Tobias Büttner <dev@spiritcroc.de>

 
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Tested devices

- [x] Physical
- [ ] Emulator
- OS version(s): Android 11

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

**Comment by ganfra** [comment]
Thats weird you got multiple same chunk in the db with only one valid...

**Comment by SpiritCroc** [comment]
This seems to happen more often though, we got a few user reports of similar endless loading issues. With the first commit, such issues still happened, but got fixed after re-opening the room. Since the second commit, I haven't observed any new reports that sound like this issue.

**Comment by ganfra** [comment]
Thanks, I'll check a bit! 
Are you working at Beeper now? :D

**Comment by SpiritCroc** [comment]
Yes, (part-time) since around April, Beeper-Android was technically a fork of SchildiChat initially :)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TokenChunkEventPersistor.kt [hunk 1]
```diff
@@ -80,7 +81,26 @@ internal class TokenChunkEventPersistor @Inject constructor(
 
                     val existingChunk = ChunkEntity.find(realm, roomId, prevToken = prevToken, nextToken = nextToken)
                     if (existingChunk != null) {
-                        Timber.v("This chunk is already in the db, returns")
+                        Timber.v("This chunk is already in the db, checking if this might be caused by broken links")
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TokenChunkEventPersistor.kt:84]: Can you just create a method to keep this a bit smaller

Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TokenChunkEventPersistor.kt:84]: By the way, is this still necessary with your second commit?

Comment (SpiritCroc) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TokenChunkEventPersistor.kt:84]: I can probably do that this weekend or next week.
It's probably only needed if we want to retrospectively fix existing cases of the issue without requiring users to do a fresh initial sync to fix it. Commit two should avoid this scenario, if I haven't missed any case.

Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TokenChunkEventPersistor.kt:84]: Yeah, this should, even if there is probably another issue underneath as I don't think we should have multiple same chunks in the db :/

Comment (SpiritCroc) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TokenChunkEventPersistor.kt:84]: We still have some issues with timeline chunk loops, like in this issue: https://github.com/vector-im/element-android/issues/5092. With some added debugging, I've seen loops starting with 4 chunks, and up to 81 chunks forming a loop (timeline chunks linking forwards and backwards indirectly to the same chunk). Seems more likely to happen while viewing a room in which spam is going on. I wonder if those loops are caused by a related bug to this one.

Comment (ganfra) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TokenChunkEventPersistor.kt:84]: it's kind of hard to replicate, but there is something wrong definitively. 
Nevermind, let's merge this to mitigate. Thanks!

**Review by ganfra — APPROVED** [review]

**Pull request merged**